File: src/main/java/io/github/oleksivio/telegram/bot/core/controller/annotations/parser/filter/telegram/PassportFileFilterParser.java
package io.github.oleksivio.telegram.bot.core.controller.annotations.parser.filter.telegram;

import org.springframework.stereotype.Component;
import io.github.oleksivio.telegram.bot.api.annotations.filter.primitive.IntegerFilter;
import io.github.oleksivio.telegram.bot.api.annotations.filter.telegram.PassportFileFilter;
import io.github.oleksivio.telegram.bot.api.model.objects.passport.PassportFile;
import io.github.oleksivio.telegram.bot.core.controller.annotations.parser.filter.FilterParser;
import io.github.oleksivio.telegram.bot.core.controller.annotations.parser.finder.Finder;
import io.github.oleksivio.telegram.bot.core.controller.handler.check.Validator;
import io.github.oleksivio.telegram.bot.core.controller.handler.check.impl.UnionExtractValidator;

import java.util.Optional;

@Component
public class PassportFileFilterParser implements FilterParser<PassportFileFilter, PassportFile> {

    @Override
    public Validator<PassportFile> createChecker(PassportFileFilter annotation, Finder finder) {
        UnionExtractValidator<PassportFile> unionExtractValidator = new UnionExtractValidator<>();

        IntegerFilter fileSize = annotation.fileSize();
        if (fileSize.status().isActive()) {
            unionExtractValidator.add(in -> Optional.ofNullable(in.getFileSize()), finder.find(fileSize));
        }
        IntegerFilter fileDate = annotation.fileDate();
        if (fileDate.status().isActive()) {
            unionExtractValidator.add(in -> Optional.ofNullable(in.getFileDate()), finder.find(fileDate));
        }

        return unionExtractValidator;
    }

    @Override
    public Class<PassportFileFilter> getAnnotationClass() {
        return PassportFileFilter.class;
    }

}

File: src/main/kotlin/io/github/oleksivio/telegram/bot/core/controller/annotations/parser/filter/telegram/PassportFileFilterParser.kt
package io.github.oleksivio.telegram.bot.core.controller.annotations.parser.filter.telegram

import io.github.oleksivio.telegram.bot.api.annotations.filter.composite.PassportFileFilter
import io.github.oleksivio.tl.kbot.server.api.objects.passport.PassportFile
import io.github.oleksivio.telegram.bot.core.controller.annotations.parser.filter.FilterParser
import io.github.oleksivio.telegram.bot.core.controller.annotations.parser.finder.Finder
import io.github.oleksivio.telegram.bot.core.controller.handler.check.Validator
import io.github.oleksivio.telegram.bot.core.controller.handler.check.impl.UnionExtractValidator
import org.springframework.stereotype.Component

@Component
class PassportFileFilterParser : FilterParser<PassportFileFilter, PassportFile> {

    override val parserAnnotationClass = PassportFileFilter::class

    override fun createChecker(annotation: PassportFileFilter, finder: Finder): Validator<PassportFile> {
        val unionExtractValidator = UnionExtractValidator<PassportFile>()

        annotation.validator
                .map { validatorName -> finder.find(validatorName, PassportFile::class) }
                .forEach { validator -> unionExtractValidator.add({ it }, validator) }

        val fileSize = annotation.fileSize
        if (fileSize.status.isActive) {
            unionExtractValidator.add({ it.fileSize }, finder.find(fileSize))
        }
        val fileDate = annotation.fileDate
        if (fileDate.status.isActive) {
            unionExtractValidator.add({ it.fileDate }, finder.find(fileDate))
        }

        return unionExtractValidator
    }

}

